File: sonnet-java/app/src/main/java/org/example/parser/Parser.java
package org.example.parser;

import java.util.ArrayList;
import java.util.List;
import org.example.syntax.AST;
import org.example.token.Token;

public class Parser {
	public Parser() {}

	public AST parse(List<Token> tokenStream) {
		if (tokenStream.isEmpty()) {
			return new AST.ListNode(List.of());
		}

		Token tok = tokenStream.removeFirst();

		return switch (tok) {
			case Token.IntToken(var val) -> new AST.IntNode(val);
			case Token.FloatToken(var val) -> new AST.FloatNode(val);
			case Token.StringToken(var val) -> new AST.StringNode(val);
			case Token.IdentifierToken(var val) -> {
				if (val.equals(":")) {
					AST quotedContent = parse(tokenStream);
					yield new AST.ListNode(List.of(new AST.IdentNode("quote"), quotedContent));
				} else {
					yield new AST.IdentNode(val);
				}
			}
			case Token.OpeningParen() -> {
				ArrayList<AST> listChildren = new ArrayList<>();
				while (!tokenStream.isEmpty() && !(tokenStream.getFirst() instanceof Token.ClosingParen)) {
					listChildren.add(parse(tokenStream));
				}
				if (tokenStream.isEmpty()) {
					throw new ParserError();
				}
				tokenStream.removeFirst();
				yield new AST.ListNode(listChildren);
			}
			case Token.ClosingParen() -> throw new ParserError();
		};
	}
}

File: sonnet-scala/src/parser/Parser.scala
package parser

import syntax.Node
import token.Token
import scala.collection.mutable.ListBuffer

class ParserError extends RuntimeException("Parser error encountered")

class Parser {

    def parse(tokenStream: List[Token]): Node = {
        if (tokenStream.isEmpty) {
            Node.ListNode(List.empty)
        } else {
            val (resultNode, remainingTokens) = parseNext(tokenStream)
            resultNode
        }
    }

    private def parseNext(tokens: List[Token]): (Node, List[Token]) = {
        tokens match {
            case Nil =>
                (Node.ListNode(List.empty), Nil)

            case Token.IntToken(value) :: tail =>
                (Node.IntNode(value), tail)

            case Token.DoubleToken(value) :: tail =>
                (Node.DoubleNode(value), tail)

            case Token.StringToken(value) :: tail =>
                (Node.StringNode(value), tail)

            case Token.IdentifierToken(name) :: tail =>
                (Node.IdentNode(name), tail)

            case Token.QuoteToken() :: tail => {
                val (quotedNode, rest) = parseNext(tail)
                val listNode           = Node.ListNode(List(Node.IdentNode("quote"), quotedNode))
                (listNode, rest)
            }

            case Token.OpeningParen() :: tail => {
                val listChildren  = ListBuffer[Node]()
                var currentTokens = tail

                while (currentTokens.nonEmpty && !currentTokens.head.isInstanceOf[Token.ClosingParen]) {
                    val (childNode, nextTokens) = parseNext(currentTokens)
                    listChildren += childNode
                    currentTokens = nextTokens
                }

                currentTokens match {
                    case Token.ClosingParen() :: rest =>
                        (Node.ListNode(listChildren.toList), rest)
                    case _ =>
                        throw new ParserError()
                }
            }

            case Token.ClosingParen() :: _ =>
                throw new ParserError()
        }
    }
}
